internal/compiler/generator/rust.rs

```diff
@@ -2849,13 +2849,43 @@ fn compile_builtin_function_call(
                         });
                     )
                 };
-                let fw_sub_menu = forward_callback(access_sub_menu, quote!(sub_menu));
-                let fw_activated = forward_callback(access_activated, quote!(activated));
+                let fw_sub_menu = forward_callback(access_sub_menu.clone(), quote!(sub_menu));
+                let fw_activated = forward_callback(access_activated.clone(), quote!(activated));
                 quote! {
                     let entries = #entries;
+
+                    let context_menu_item_tree = if sp::WindowInner::from_pub(#window_adapter_tokens.window()).supports_native_menu_bar() {
+                        // May seem overkill to have an instance of the struct for each call, but there should only be one call per component anyway
+                        struct ContextMenuWrapper(sp::VRc<sp::ItemTreeVTable, #popup_id>, sp::ModelRc<sp::MenuEntry>);
+                        const _ : () = {
+                            use slint::private_unstable_api::re_exports::*;
+                            MenuVTable_static!(static VT for ContextMenuWrapper);
+                        };
+                        impl sp::Menu for ContextMenuWrapper {
+                            fn sub_menu(&self, parent: sp::Option<&sp::MenuEntry>, result: &mut sp::SharedVector<sp::MenuEntry>) {
+                                let self_rc = self.0.clone();
+                                let _self = self_rc.as_pin_ref();
+                                let model = match parent {
+                                    None => self.1.clone(),
+                                    Some(parent) => #access_sub_menu.call(&(parent.clone(),))
+                                };
+                                *result = model.iter().map(|v| v.try_into().unwrap()).collect();
+                            }
+                            fn activate(&self, entry: &sp::MenuEntry) {
+                                let self_rc = self.0.clone();
+                                let _self = self_rc.as_pin_ref();
+                                #access_activated.call(&(entry.clone(),));
+                            }
+                        }
+                        Some(sp::VBox::new(ContextMenuWrapper(popup_instance.clone(), entries.clone())))
+                    }
+                    else {
+                        None
+                    };
+
                     {
```

Comment:
I think we only need this part when not using the native context menu (fallback)